### PR TITLE
KNL-1268 - Extract content file operations to support multiple backends

### PR DESCRIPTION
**Pull request by botimer**
Thanks to OpenCollab ZA for the patch.

This makes way for filesystems other than local (or mapped) disk, such
as OpenStack Swift.

A new interface is added, FileSystemHandler, along with extracting the
default file implementation. I made a few fixes to update the patch and
clean up a few stylistic things to fit into the kernel code better, but
there are no name or design changes.

All tests are passing and I have verified that DB and the local file backend are working as usual. The next step will be to test out the alternate backend implementation for Swift and verify if OpenCollab has made any adjustments since this patch was supplied.

### CODE REVIEW

**Comment by ottenhoff** [comment]
Looks good.  Can you change the author on the commit to OpenCollab?  Then you become the committer and OpenCollab would be the patch author.

**Comment by botimer** [comment]
Is it clear who the author should be as far as name/email?

**Comment by ottenhoff** [comment]
Fair point, it appears Jaques Smith does not have a Github account.... okay, let's proceed as is!

**Comment by buckett** [comment]
You can make someone the author without them having a github account.

On 13 April 2015 at 15:26, Sam Ottenhoff <notifications@github.com> wrote:

> Fair point, it appears Jaques Smith does not have a Github account....
> okay, let's proceed as is!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sakaiproject/sakai/pull/394#issuecomment-92377997>.
>



-- 
  Matthew Buckett

File: kernel/api/src/main/java/org/sakaiproject/content/api/FileSystemHandler.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    public InputStream getInputStream(String id, String root, String filePath) throws IOException;
+
+    /**
+     * Save the file from the input stream to the path and return the content size.
+     * 
+     * @param id The id of the resource. Will not be null or empty.
+     * @param root The root of the storage. Could be null or empty.
+     * @param filePath The path to save the file to. Will not be null or empty.
+     * @param stream The stream to read the file from.
```
Comment (buckett) [inline kernel/api/src/main/java/org/sakaiproject/content/api/FileSystemHandler.java]: Who is responsible for closing the stream? The caller of the method or `saveInputStream()` itself? From the default implementation it looks like the caller is.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/ContentServiceSqlDefault.java [hunk 2]
```diff
@@ -123,6 +123,14 @@ public String getResourceIdXmlSql()
 		return "select RESOURCE_ID, XML, BINARY_ENTITY from CONTENT_RESOURCE where FILE_PATH IS NULL";
 	}
 
+    /**
```
Comment (buckett) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/ContentServiceSqlDefault.java]: Indentation doesn't look the same as the surrounding lines.

**Comment by buckett** [comment]
My 2 comments aren't show stoppers, just things I noticed while reading. Thanks for this and it would be good to get it in.

File: kernel/api/src/main/java/org/sakaiproject/content/api/FileSystemHandler.java (hunk 1 above)
Comment (botimer) [inline kernel/api/src/main/java/org/sakaiproject/content/api/FileSystemHandler.java]: Yes, this seems right to me. Unless we want to do some sort of transactional callback interface, the caller will be responsible for cleaning up. This is probably expectable and reasonable.

**Comment by botimer** [comment]
I've pushed a new version of this commit with the author set to Jaques and the whitespace fixed. 
I'd say this is ready to merge.

(I can't believe I got caught on that one, given the number of the other places cleaned up...)

File: kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/util/StorageConverter.java [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        print("Properties (optional):");
+        print("- deleteFromSource: Whether to delete the source files. Default false.");
+        print("- contentSql: The sql statement to retrieve the resource id's and paths. Default is new ContentServiceSqlDefault().getResourceIdAndFilePath()");
+        print("----------------------------------------------------------------------");
+    }
+
+    /**
+     * Print the text to the screen.
+     */
+    private static void print(String text){
```
Comment (ern) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/util/StorageConverter.java]: In StorageConverter.main there is a mix of using System.out.println and the use of a logger?

**Comment by ottenhoff** [comment]
@botimer can you review latest comments?

**Comment by botimer** [comment]
Sure thing @ottenhoff.

The StorageConverter is only used from the exec plugin, so I saw no problem with System.out there. The messages go to the interactive shell, which is appropriate.

In general, this is in good enough shape to be merged. I have run into some tricky classloader/Maven/Spring issues when plugging in alternate implementations, but those should not hold up this merge. I will have lots of time this weekend to nail down the loose ends of the Swift/BlobStore (generalized object store, S3) handler. It looks like the best (and easiest overall) way to handle this is to do something similar to how the Profile manager was changed to a proxy -- not the most elegant but pretty practical and configurable by name from xyz.properties.

**Pull request merged**